### PR TITLE
feat: Tenant specific config support via .cac.toml

### PR DESCRIPTION
**Pull request by ayushjain17**
## Problem
Tenant specific configuration not present

## Solution
Get tenant-config via cac using `cac_toml` which reads `Superposition.cac.toml`

### CODE REVIEW

File: crates/cac_toml/Cargo.toml [hunk 1]
```diff
@@ -17,4 +17,4 @@ path = "src/bin.rs"
 clap = { version = "4.3.0", features = ["derive"] }
 pest = "2.6"
 pest_derive = "2.6"
-toml = { version = "0.8.8", features = ["preserve_order"] }
+toml = { workspace = true }
```
Comment (knutties) [inline crates/cac_toml/Cargo.toml:20]: Will this pick-up the `preserve_order` if someone just includes `cac_toml` outside of this workspace?  Could we validate that?

Comment (ayushjain17) [reply crates/cac_toml/Cargo.toml:20]: yes this will set the features for the crate, in the same way it is specified in the workspace `Cargo.toml`
have verified this

File: crates/superposition_types/src/lib.rs [hunk 2]
```diff
@@ -257,6 +257,46 @@ impl Display for RegexEnum {
     }
 }
 
+pub trait SuperpositionTenantConfig {
```
Comment (Datron) [inline crates/superposition_types/src/lib.rs:260]: This could be a struct with pub members right? Why define tenant config as a trait behaviour? No one can implement this elsewhere and pass it around. Why are we making it generic? Note: If the answer summarizes to "We will need this in the future", my response is "lets add it in the future". This is unecessary code complexity right now

File: crates/superposition/src/main.rs [hunk 3]
```diff
@@ -64,69 +48,16 @@ async fn main() -> Result<()> {
         prefix => "/".to_owned() + prefix,
     };
 
-    let cac_host: String = get_from_env_unsafe("CAC_HOST").expect("CAC host is not set");
     let cac_port: u16 = get_from_env_unsafe("PORT").unwrap_or(8080);
```
Comment (Datron) [inline crates/superposition/src/main.rs:51]: Nice, we needed this

**Review by Datron — CHANGES_REQUESTED** [review]
Let me know if we can avoid the type ceremony regarding TenantConfig being a trait. Else this is fine

File: crates/superposition_types/src/lib.rs (hunk 2 above)
Comment (Datron) [reply crates/superposition_types/src/lib.rs:260]: I'm guessing this is because of internal asks for auth? The tenant config struct internal to Juspay will be different I'm assuming

**Review by Datron — APPROVED** [review]

**Review by Datron — APPROVED** [review]

File: crates/superposition/src/main.rs [hunk 4]
```diff
@@ -153,51 +89,39 @@ async fn main() -> Result<()> {
...
         let leptos_options = &conf.leptos_options;
         let site_root = &leptos_options.site_root;
         let leptos_envs = ui_envs.clone();
-        let cac_host = cac_host.to_owned() + base.as_str();
         App::new()
             .wrap(Compress::default())
+            .app_data(app_state.clone())
             .wrap_fn(|req, srv| {
-                let user = User::default();
+                let state = req.app_data::<Data<AppState>>().unwrap();
```
Comment (ShubhranshuSanjeev) [inline crates/superposition/src/main.rs:103]: @ayushjain17 , can you explain why this change is needed?

File: crates/experimentation_platform/src/api/experiments/handlers.rs [hunk 5]
```diff
@@ -234,30 +244,48 @@ async fn create(
     // creating variants' context in CAC
     let http_client = reqwest::Client::new();
     let url = state.cac_host.clone() + "/context/bulk-operations";
-    let headers_map = construct_header_map(tenant.as_str(), custom_headers.config_tags)?;
+    let user_str = serde_json::to_string(&user).map_err(|err| {
```
Comment (ShubhranshuSanjeev) [inline crates/experimentation_platform/src/api/experiments/handlers.rs:247]: What are we trying to achieve with `x-user` header?

**Review by ShubhranshuSanjeev — CHANGES_REQUESTED** [review]

File: crates/experimentation_platform/src/api/experiments/handlers.rs (hunk 5 above)
Comment (ayushjain17) [reply crates/experimentation_platform/src/api/experiments/handlers.rs:247]: Someone who has access for only making changes in experimentation and not in context-aware-config, for them all requests of experimenation would start failing as the when these api calls would be made to make changes in context-aware-config, the authorization would fail due to lack of access.

So Internal authorization has been added
Now since this Internal auth would not have the details of the current user, and on validating it would give the details of a super-user then the audit logs would show wrong info regarding the maker of the contexts in context-aware-config, so in this case since experimentation has already decoded the user, it just forwards the user data 

And this user data is only to be consumed if the Internal auth check is successful

**Review by ShubhranshuSanjeev — APPROVED** [review]

File: crates/experimentation_platform/src/api/experiments/handlers.rs [hunk 6]
```diff
@@ -794,29 +841,47 @@ async fn update_overrides(
...
+    let headers_map = construct_header_map(tenant.as_str(), extra_headers)?;
 
     let response = http_client
         .put(&url)
         .headers(headers_map.into())
         .header(
-            "Authorization",
-            format!("{} {}", user.get_auth_type(), user.get_auth_token()),
+            header::AUTHORIZATION,
+            format!("Internal {}", state.superposition_token),
```
Comment (Datron) [inline crates/experimentation_platform/src/api/experiments/handlers.rs:866]: What is internal auth?

**Review by Datron — CHANGES_REQUESTED** [review]
want to understand Authorization internal before approving

File: crates/experimentation_platform/src/api/experiments/handlers.rs (hunk 6 above)
Comment (Datron) [reply crates/experimentation_platform/src/api/experiments/handlers.rs:866]: This is not a supported authentication scheme
https://developer.mozilla.org/en-US/docs/Web/HTTP/Authentication#authentication_schemes

**Review by ShubhranshuSanjeev — CHANGES_REQUESTED** [review]
Can we restrict the internal token access to be only used by experimentation platform and also that this token should not have the full access, just access to CRUD of contexts/overrides will work?

File: crates/experimentation_platform/src/api/experiments/handlers.rs (hunk 6 above)
Comment (Datron) [reply crates/experimentation_platform/src/api/experiments/handlers.rs:866]: https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml

File: crates/experimentation_platform/src/api/experiments/handlers.rs [hunk 7]
```diff
@@ -234,30 +244,48 @@ async fn create(
...
+        .into_iter()
+        .filter_map(|item| match item {
+            ContextBulkResponse::PUT(context) => Some(context),
+            ContextBulkResponse::DELETE(_) => None,
+            _ => {
+                log::error!("Unexpected response item: {:?}", item);
+                None
+            }
+        })
+        .collect::<Vec<_>>();
```
Comment (Datron) [inline crates/experimentation_platform/src/api/experiments/handlers.rs:288]: nitpick: specify types here, makes it easy to read on a glance

File: crates/experimentation_platform/src/api/experiments/handlers.rs [hunk 8]
```diff
@@ -371,6 +399,16 @@ pub async fn conclude(
                 };
                 operations.push(ContextAction::MOVE((context_id, context_move_req)));
             } else {
+                let user_str = serde_json::to_string(&user).map_err(|err| {
```
Comment (Datron) [inline crates/experimentation_platform/src/api/experiments/handlers.rs:402]: Do we want to send the whole user object in the headers or only the email address?

**Review by Datron — APPROVED** [review]
Minor nitpicks and optimizations, not a PR blocker

File: crates/experimentation_platform/src/api/experiments/handlers.rs (hunk 8 above)
Comment (ayushjain17) [reply crates/experimentation_platform/src/api/experiments/handlers.rs:400]: Whole user, as what all data would be needed later on the context api or which api being called would not be known

**Review by ShubhranshuSanjeev — APPROVED** [review]
